### DISCUSSION TITLE
On the project's MVP

### DISCUSSION THREAD

**Discussion by Philip-Clark** [club-40/odin-pantry#20] [General]
Let's start discussing the base requirements we need to meet in order to consider this project complete.

Below is the project as it was submitted:

> ### Pantry/Fridge manager
> I'll volunteer the idea I had for my final project. A web app that let's you search and add items you have placed in your fridge and pantry with timers on when they are set to expire. Let's you track your grocery needs list and share it with users. Would have a backend and require oauth and could have multiple users share a single 'fridge' There's a barcode and product lookup api I have saved for this, but that would have been for mobile to use the camera

~~Please see the MVP section of the [project's Google document](https://docs.google.com/document/d/1jvajnjTTG5G4mqY-7VTETwCf_YJRjIbwEMTONbQkjB8/edit) if you are looking for a list of MVP features.~~

Please see the MVP section of the [projects overview document](https://github.com/club-40/odin-pantry/blob/main/docs/project-overview.md#features--mvp-) if you are looking for a list of MVP features.

**Comment by Philip-Clark** [comment]
I think the following is a good set of features to start us out.

1. Add/remove items from the pantry.
2. Multiple cupboards or pantry sections. eg: The fridge, the main pantry, the small pantry
3. Items should have a name, entrance date, expiration date, and the option to remove.
4. There should be an option for multiple users, with items being assigned an owner.
5. Users should be able to sign in via username/password.
6. A backend to store all users' items.

This is a non-exhaustive list, and just a few ideas I think would make for a solid, but achievable MVP.

**Reply by JustWaveThings** [reply]
number 3 --- consumed date? Or somehow to remove the item from the inventory or an update with a new date .

**Reply by Philip-Clark** [reply]
> @JustWaveThings,
> number 3 --- consumed date? Or somehow to remove the item from the inventory or an update with a new date .

I think number 1 covers that, but I updated the list to be more clear.

**Comment by JustWaveThings** [comment]
Here is an api  we can use to pull the product info from from like 20 chains. I think we might be able to pull images from it.

 --- https://developer.kroger.com/

>All Public APIs provide access to the following Kroger subsidiaries:

> Kroger
> Bakers
> City Market
> Dillons
> Food 4 Less
> Foods Co
> Fred Meyer
> Fry's
> Gerbes
> JayC
> King Soopers
> Metro Market
> Owen's
> Pay Less
> Pick 'n Save
> QFC
> Ralphs
> Smith's
>

**Comment by JustWaveThings** [comment]
Here is an API  called still-tasty  --- https://stilltasty.com/

<img width="813" alt="image" src="https://user-images.githubusercontent.com/48169481/230507321-9794862e-616f-43c5-a76e-bbaddeb7bce2.png">

**Reply by JustWaveThings** [reply]
Note that I am not married to any specific API's -- just these were two that I found with a quick search

**Reply by Ananth-Adhikarla** [reply]
they probably are the only free and open source ones that are available so good find.

**Reply by taladan** [reply]
Hey, looking at stilltasty - I couldn't find anything on their site regarding API documentation.  Is it an actual API, or would we have to scrape their site to query their DB?

**Comment by justindanielclark** [comment]
No clue how helpful this will be at this stage, but I had drafted the following schema a while ago which is fairly in line with your initial MVP, Biff. (It's the pdf at the bottom in the link)

I always found drafting out the relations between data helped me conceptualize what the app is and what it needs to be able to do.

Users are made up of their email and a password. We will need a way for users to log in.

Products, in the vague sense, are both a product archetype (the idea of a banana, for instance) and a product in actuality (the banana on your counter). Users create archetypes so they can reference them again quickly in lookups when adding things they have already added once before.

The archetype has a base product_type, which is an defined item from an enum list (to help with sorting) and a name. Users will tend to buy/reference many of the same archetype items over and over again, so having each user store how they refer to each archetype seemed to make sense to me. This eliminates any need for this app to handle what a banana is in every possible country code or for the app to pre-conceive of every possible product that could be added. It does this at the cost of having to track N items types for every M users.

The product in reality has a created date, an expectation when it expires, and an optional quantity. A user could have a banana they bought on Tuesday and 3 bananas they bought on Friday. Theres no reason to track 4 bananas, but two separate banana entries.

A pantry is made up of product in reality entries. Every pantry has an id

The combination of the pantry id and the user id creates a pantry subscription, which lets us lookup which pantry has which users and which users belong to which pantries.

Users should be able to:
- Log In / Log Out
- Create a pantry
- Join a pantry
- Unsubscribe from a pantry
- Send out a request to another to join an existing pantry. If that other is an existing member, it should be reflected in the app. Otherwise, an email I guess
- Create product archetypes for storage for later.
- Add a product that conforms to an archetype to a pantry
- Reduce the quantity of a product in a pantry
- Delete a product entry from a pantry

[edit - added the below later]

- When adding a product entry, the user should be prompted to 'lookup' if they have added that product before. If not, it should allow them to create a new productArchetype.
- Archetypes should have a user defined expected expiration timer. A user who lives in Finland in the winter who puts their bananas in the fridge has them exprire at a different rate than a user in Chile who leaves them on the counter every time.
- It should have views that break down foods by category. A user may want to see what fruits they have in the house, if any.
- Perhaps, products could have a range of times that define its quality. I know bananas are good for about 5 days in my house. I really enjoy eating them typically days 1-2. Days 3-5 are less than great. Day 5+ it's a mushy mess. Users can opt into their archetypes supporting a greatBy, goodBy, and expiredBy date timer which is reflected in status.
- Product entries should support locations which would look like a tag, I guess?
- Locations should be user defined (fridge, pantry, counter, ...) the same way ProductArchetypes are

[roughSchema.pdf](https://github.com/Club-40/group-project-1/files/11175163/roughSchema.pdf)

**Reply by justindanielclark** [reply]
Also, if it is helpful, I drafted this relational map using this website [https://dbdiagram.io/] and this data-set:

Table users {
  id int [pk, increment]
  email string [not null]
  password string [not null, note: 'hashed']
}

Table pantrys {
  id int [pk, increment]
}

Table pantry_subscription {
  id int [pk]
  pantry_id int [ref: > pantrys.id]
  user_id int [ref: > users.id]
}

Table productArchetypes {
  id int [pk, increment]
  user_id int [ref: > users.id]
  name string [not null]
  type product_types [not null]
  refreshed Date [not null, note: 'a date that represents the last time this archetype was used or referenced']
}

Table products {
  id int [pk, increment]
  pantry int [ref: > pantrys.id]
  archetype_id int [ref: > productArchetypes.id]
  created Date [not null]
  expires Date [not null]
  quantity int
}

Enum product_types {
  //Fruits, Veg, Nuts
  vegetable
  fruit
  nut
  //Meat
  meat
  beef
  pork
  chicken
  fish
  lunchmeat
  imitation
  tofu
  //Grains and Bread
  grain
  bread
  pasta
  //Oils and Spices
  oil
  vinegar
  spice
  condiment
  //Dairy
  dairy
  milk
  egg
  cheese
  yoghurt
  icecream
  butter
  //Drinks
  drink
  beer
  alcohol
  wine
  //Processed
  snack
  dessert
}

**Reply by JoshDevHub** [reply]
You can diagram ERs with [mermaid](https://mermaid.js.org/syntax/entityRelationshipDiagram.html) by the way (markdown-ish diagram syntax supported by GitHub)

~~~mermaid
erDiagram
    USER ||--o{ PANTRY_SUBSCRIPTION : has_many
    USER {
        int id PK
        string email "not_null"
        string password "not_null"
    }
    PANTRY ||--o{ PANTRY_SUBSCRIPTION : has_many
    PANTRY ||--o{ PRODUCT : has_many
    PANTRY {
        int id PK
    }
    PRODUCT ||--|| PRODUCT_ARCHETYPE : has_one
    PRODUCT {
        int id PK
        int pantry_id FK
        int archetype_id FK
        int quantity
        date created
        date expires
    }
    PRODUCT_ARCHETYPE {
        int id PK
        string name
        type product_type
        date refreshed
    }
    PANTRY_SUBSCRIPTION {
        int id PK
        int pantry_id FK
        int user_id FK
    }
~~~

(absent the big enum and most of the null constraints because I'm lazy)

There are a few questions I might have here. I think I get the idea with `product_archetypes`, but why does it need a user ref?

Anyways, I wouldn't be inclined to say that getting the data modeling hammered down is in the scope of discussing MVP. We mostly want to be focused on features -- think very high level, what can the user do on the site?

**Reply by Mogoatlhe** [reply]
i could be wrong, but i think based on the description above the user ref in the `product_archetypes` is so that if that specific user created the archetype they can then references it in the future, i.e. users who haven't had that archetype won't haven't it show up for them?
or it's just a mistake lol.

but i think it's better to have it available for all users once it's created by one, so no user ref needed. that is if i understood what i read correctly.

**Reply by justindanielclark** [reply]
Mogoalthe has it right on the money. The idea was that a user only is shown the archetypes they have created before when looking to add an item into their pantry.

Also, I do think you are right that discussion data structure doesn't fit MVP exactly but I do think talking about data and how it ought be structured does illuminate functionality that  you would want.

**Reply by justindanielclark** [reply]
Also, the ER modeling directly in GH is awesome, thanks for sharing that. Bookmarked.

**Reply by justindanielclark** [reply]
A few more features that the app should support:
- When adding a product entry, the user should be prompted to 'lookup' if they have added that product before. If not, it should allow them to create a new productArchetype. 
- Archetypes should have a user defined expected expiration timer. A user who lives in Finland in the winter who puts their bananas in the fridge has them exprire at a different rate than a user in Chile who leaves them on the counter every time.
- It should have views that break down foods by category. A user may want to see what fruits they have in the house, if any.
- Perhaps, products could have a range of times that define its quality. I know bananas are good for about 5 days in my house. I really enjoy eating them typically days 1-2. Days 3-5 are less than great. Day 5+ it's a mushy mess. Users can opt into their archetypes supporting a greatBy, goodBy, and expiredBy date timer which is reflected in status.
- Product entries should support locations which would look like a tag, I guess?
- Locations should be user defined (fridge, pantry, counter, ...) the same way ProductArchetypes are

**Reply by mmxz-mmxz** [reply]
I think product/pantry archetypes are a bit out of scope for a MVP, but this is a good start.  
  
My suggestions (not mentioning auth as how this is being handled is still being decided):  
  
Users should be able to:
* Create a pantry  
  * `pantry/new` - accepts the following fields:  
    * `name`  
    * `description` (optional)  
* **Delete a pantry**  
  * `pantry/:id/delete`  
    * Should throw an error if user is not owner of pantry, or if pantry with passed id does not exist  
    * Should unsubscribe all users subscribed to pantry  
* Join (subscribe to) a pantry  
  * `pantry/:id/invite` - accepts the following fields:  
    * `userId` - id of user being invited to pantry  
    * Should throw an error if user is not the owner of the pantry  
  * `pantry/:id/join`  
    * Should throw an error if an active invite for user does not exist or pantry does not exist  
    * Should delete active invitation for user  
* Leave (unsubscribe) from a pantry  
  * `pantry/:id/leave`  
    * Should throw an error if user is not a member of the pantry or pantry does not exist  
* Send out a request to another to join an existing pantry. If that other is an existing member, it should be reflected in the app. Otherwise, an email I guess  
  * `pantry/:id/invite` should send email to invited user  
* ~~Create product archetypes for storage for later.~~  
  * Out of scope
* ~~Add a product that conforms to an archetype to a pantry~~  
  * Out of scope  
* View all pantries created/subscribed to  
  * `user/:id/pantries`  
    * Should throw an error if requesting user id is not the same as passed id  
    * Should list all created and subscribed to pantries (? - can also have separate route for subscribed pantries)  
* View all items in a pantry they are a member of  
  * `pantry/:id/items`  
  * Should throw an error if user is not owner of or subscribed to pantry  
* Add an item to a pantry  
  * `pantry/:id/add` - should accept the following fields:  
    * `id` - id of item being added  
    * `quantity` - quantity of item being added  
    * `expires`  - user should be able to add expiration, not sure if barcode stuff is MVP  
    * Should throw error if user is not owner of pantry or item does not exist  
    * `id` optional in favor of `name`? Not sure how we want to handle product lookup in FE  
* Reduce the quantity of a product in a pantry  
  * `pantry/:id/updateItem` - should accept the following fields:  
    * `id` - id of item being updated  
    * `quantity` (optional)  
    * `expires` (optional)  
    * Should throw error if user is not owner of pantry, pantry does not exist, item does not exist  
* Delete a product entry from a pantry  
  * `pantry/:id/removeItem` - should accept the following fields:  
    * `id` - id of item being removed  
    * Should throw error is user is not owner of pantry, pantry does not exist, item does not exist

**Reply by Zadag** [reply]
Just to clarify on the product_archetype table; is the idea that that would be used to categorize different types of products?  Like Dairy, Meat, Vegetables etc.  Would these archetypes be user defined?

**Reply by Ananth-Adhikarla** [reply]
As discussed on Discord, its better to start with the list of features first then move on to making the database and other structures.

I have made a google doc with the project description and other things to start out with. It has editable access, if it doesn't work let me know or you can make a copy of it.

Contributor list was made based on the introduction in GitHub, if it doesn't have your name of if your prefer any other name let me know as well.
 
https://docs.google.com/document/d/1jvajnjTTG5G4mqY-7VTETwCf_YJRjIbwEMTONbQkjB8/edit?usp=sharing

**Reply by Cedrus32** [reply]
> https://docs.google.com/document/d/1jvajnjTTG5G4mqY-7VTETwCf_YJRjIbwEMTONbQkjB8/edit?usp=sharing

Ananth, this is awesome! Thanks for taking the time to do this.

Some of the items under MVP Features feel advanced and better suited to a later version (recipes, barcode scanner, estimated best-by dates, budgeting calculator, timeline for purchases and spending). These are all really cool features, I think could be better addressed later when we have our minimal "bare bones" product running smoothly. I don't want to change anything there without consensus, so this is me starting the discussion.

**Reply by Ananth-Adhikarla** [reply]
Hey Cedrus

This is just a starting template which is collecting raw ideas for now for the features, once everyone agrees to what should be included and not then it will be revised for the project.

You are free to change or make a copy of it and re write the ideas as you wish or add comments where required.

**Comment by JoshDevHub** [comment]
Should users also be able to edit/update items in their pantry? I think it probably makes sense to allow, partially just for the good user experience of fixing a small mistake without having to delete the item + remake a new one. Full CRUD makes sense imo.

Can users move food from one of their "pantry" sections to a different one (ie. someone moving meat from the fridge to the freezer)? And if so, how do we handle expiration changing in different contexts? Using that same example, the meat's lifetime is of course significantly extended in the freezer. This could be a large enough problem that we may want to kick it down the road past MVP, but imo it's worth thinking about.

RE: Authentication -- do we want our own auth system or do we want to use OAuth (and via which providers) or do we want to implement both? I'm personally in favor of OAuth and could probably take or leave having our own system in place (glad to expand on this).

Are we interested in contacting the users by email at all (ie. for a welcome email on signing up)? Are we interested in having our app receive emails?

**Reply by Lofty-Brambles** [reply]
Instead of using emails, I'd rather be having an OTP system for creating shareable links for a pantry. Also, maybe having authorisation on the similar lines with 3-4 tiers like "viewers", "managers", "admins", "owners" with view-check perms, create-edit-delete perms, manage-members perms and manage-the-whole-pantry perms respectively.

**Reply by RyanMcEntire** [reply]
If we're also doing emails, email might be an opt in way of letting the user know that the milk they bought is about to expire. Since this isn't a mobile app with notifications,

**Reply by justindanielclark** [reply]
I like the idea of users being able to opt-in to receiving emails about items as they are set to expire, but perhaps the simplest experience for the user is the easiest. I personally hate the whole process of setting up an account and have left sites that hid functionality behind the process.

I am in favor of whatever system makes it as easy on the end user as possible. If the end user wants emails and the auth system we setup does not collect the data for it, just have a screen that lets them know they need to provide it to 'opt in'

**Reply by Mogoatlhe** [reply]
@JoshDevHub it does make sense for every user defined entity to be editable, for the reason you stated above.

> can users move food from one of their "pantry" sections to a different one (ie. someone moving meat from the fridge to the freezer)? And if so, how do we handle expiration changing in different contexts?

that's not even something i considered to be honest. i thought the project was going to focus explicitly on the `best before` date prescribed by the product, and we'd simply remind the user when it's close or something along those lines. with this new dynamic we'd have to decide (find out) if beef from brand A and beef from brand B get the same life expectancy when put in the freezer. i guess what i'm getting at is, will we treat all meat (of the same type, but from different providers) the same?

> RE: Authentication -- do we want our own auth system or do we want to use OAuth (and via which providers) or do we want to implement both?

given the goal of most (or some) of club 40 members' goal is employment, i think building our own might be more beneficial as something to have on CV? and i think it allows having more items on the board for people to do given the volume of interested shown for this project. admittedly  i haven't created an auth system myself or used OAuth yet, so if either one covers both points then you can ignore me.

> Are we interested in contacting the users by email at all (ie. for a welcome email on signing up)? Are we interested in having our app receive emails?

as mentioned above by @RyanMcEntire emails might be a good idea to have a way for users to opt in to receive "close to expiry" notifications, and maybe to also share pantries by email (this could be done in app though). otherwise i don't see another need for emails. and i can't think of a reason why users would want to email us, i'm assuming that's what the last question is about?

**Reply by Mogoatlhe** [reply]
@justindanielclark 
> I personally hate the whole process of setting up an account and have left sites that hid functionality behind the process.

lol, i also do this.
but wouldn't the user need to setup an account for us to persist their data? otherwise how would we know who to serve up what to? unless i misunderstood you?

**Reply by justindanielclark** [reply]
I just mean that an unwieldy process or a process with too many steps will likely turn away users. Auth that allows me to use Google or Github or whatever is way easier on the end user than the traditional email -> confirm email approach

**Reply by Mogoatlhe** [reply]
ah, i see makes sense.

**Reply by RyanMcEntire** [reply]
Yeah I also don't see a reason to hide functionality for this behind an email registration, but some people might find it helpful to be reminded about their fridge because they spend all their time on the computer.

**Comment by taladan** [comment]
Something else that we might think about wrt Pantries, do we want to allow for rotation of items, notifications of rotations and expirations, etc?  In a commercial setting, rotating out the older stuff first is typically a requirement, though I know we didn't specifically say that this app would be useful in a commercial environment, I don't know any reason why it shouldn't be able to do that sort of thing.  If we do want to add rotations, nontifications and expirations, we may want to also track things like vendors/suppliers, and maybe even add hooks for reordering supplies (either through some sort of food supplier like Sysco, or for residential pantries, ordering from their local markets online if that's an available option.

Just some thoughts - feel free to shoot them down if you think they fall outside the scope of this project.

**Reply by Mogoatlhe** [reply]
sorry if i'm being slow here, but what is meant by "rotation of items"? is it like making sure that items closest to expiry date are attended to first?

> rotating out the older stuff first is typically a requirement

based on the above, that's what i assumed you meant. i'm just confirming.

**Reply by Zadag** [reply]
> is it like making sure that items closest to expiry date are attended to first?

Exactly.  Although I think @taladan is talking about implementing notifications on a periodic basis for bulk products?

**Reply by Cedrus32** [reply]
I love this idea -- it would make the app much more useful, but feels more like it should be an added feature after we get the MVP up and running, especially if we include vendors, ordering, etc.

**Reply by Mogoatlhe** [reply]
thank you @Zadag for the clarification. i completely agree with @Cedrus32 maybe we can revisit it once we've done the mvp.

**Reply by taladan** [reply]
Yes, that is exactly what I meant.  If we can make it applicable for single users and small restaurants, it could be extremely useful. I feel like planning a good reporting structure for the data and  what kind of data we want to handle is good info to think about also.  If we know our target users, would that not inform the data and design?

**Reply by Ananth-Adhikarla** [reply]
I like this idea, very well thought about for a realistic use. To answer some of those questions

> do we want to allow for rotation of items, notifications of rotations and expirations, etc

  - Rotation of items hard to say on this but I did leave it as an option for now
  - Notifications will most likely exists but its not decided upon yet, if it exists will it be on the website itself or do you email the person 
  - Expiration is already agreed upon

> In a commercial setting, rotating out the older stuff first is typically a requirement, though I know we didn't specifically say that this app would be useful in a commercial environment, I don't know any reason why it shouldn't be able to do that sort of thing.

  - The only reason I can think of is building an app for the general public and commercial are two very different things as they both have different use cases, needs and requirements. It can be added as a feature but it would depend on the person if they want to use it or not for example. 

> If we do want to add rotations, nontifications and expirations, we may want to also track things like vendors/suppliers, and maybe even add hooks for reordering supplies

  - If we can find a good API that can give us that info then we should be good for vendor and suppliers but it can be hard to implement it and maintain. 
  - Hooks for reordering supplies, not sure what this means exactly, would this involve payment gateways and secure sessions? 
 
>  If we know our target users, would that not inform the data and design?

 - I dont think anyone has thought about this question on who the target users would be, just making the project for fun I suppose with nothing in mind.

**Comment by Philip-Clark** [comment]
MVP has been decided, please see https://github.com/club-40/odin-pantry/blob/main/docs/project-overview.md#features--mvp- for results.
